### PR TITLE
Add `/open_cfds` endpoint

### LINKED ISSUE #1711: Introduce new `open_cfds` endpoint that only returns CFDs which are considered "open"
`is_open` means we have a spendable multi-sig on the blockchain or have a transaction to create such joint output.

### PR DESCRIPTION
**Pull request by thomaseizinger**
We define "open" as "has a confirmed, unspent joint output". Our
definition of "confirmed" for these purposes is "having a transaction"
that can create such an output. This is the case after contract-setup
has finished but before we try to broadcast the transaction. Failing to
broadcast the transaction is extremely unlikely because our software is
in control of the wallet and should in normal operation prevent double-
spending.

If we ever extend our monitoring - and in a further step the event model
to account for transactions in mempool - we can potentially redefine
this as "transaction must have been successfully broadcasted".

## Note to reviewers

I set the boolean only where I deemed it necessary. Please pay close attention to whether or not it is set in all places correctly.

### CODE REVIEW

File: daemon/src/projection.rs [hunk 1]
```diff
@@ -439,6 +446,7 @@ impl Cfd {
                 self.closing_price = Some(price);
 
                 self.aggregated.state = CfdState::PendingClose;
+                self.aggregated.has_unspent_joint_output = false;
```
Comment (luckysori) [inline daemon/src/projection.rs:456]: ❓ I would assume that the `ContractSetupFailed` state is only reachable from `ContractSetupStarted`, so setting this to `false` here wouldn't be strictly necessary. Is this change deliberate so that we cover our bases in case we can go "backwards" after `ContractSetupCompleted`?

**Review by luckysori — APPROVED** [review]
LGTM! It's not immediately obvious what the state of `has_unspent_joint_output` should be after each event, but I think I agree with your choices.

File: daemon/src/projection.rs (hunk 1 above)
Comment (thomaseizinger) [reply daemon/src/projection.rs:456]: You are commenting on the `CollaborativeSettlementCompleted`. The `ContractSetupFailed` event does not set the `has_unspent_joint_output` variable.

Comment (thomaseizinger) [reply daemon/src/projection.rs:456]: GitHub is a bit mean here because it collapsed everything in-between :)

Comment (luckysori) [reply daemon/src/projection.rs:456]: Okay, that makes way more sense. Thanks for clarifying!

**Comment by thomaseizinger** [comment]
Needs to be rebased on top of https://github.com/itchysats/itchysats/pull/1717 to incorporate the new failure mode.

**Review by da-kami — CHANGES_REQUESTED** [review]
I think all "normal" cases are covered.

Only `ContractSetupCompleted` sets to true.
*Any* event that decides the closing price (or is after the price was locked in) sets to `false`.

One weird outsider is `RevokeConfirmed` - I am not sure about that one. If we find it on chain I think we should still set it to `false` for now because we won't be able to find the `attestation` / `CET`...? Obviously the price is not locked in for that one, but it is still problematic if we keep it around, because we would consider it as `Open` forever given the current logic, or am I missing something? 😬
I think it's safer to set it to `false` in that case and add a comment to change it once handling for `RevokeConfirmed` is implemented.

**Comment by klochowicz** [comment]
> Needs to be rebased on top of #1717 to incorporate the new failure mode.

rebased (applied a separate commit on top of master, so it's easier for everyone to look at the diff) - this will get squashed.

**Review by da-kami — APPROVED** [review]
thank you

File: daemon/src/projection.rs [hunk 2]
```diff
@@ -480,11 +492,13 @@ impl Cfd {
...
             }
             RefundTimelockExpired { .. } => {
                 self.aggregated.refund_published = true;
 
                 self.aggregated.state = CfdState::PendingRefund;
+                self.aggregated.has_unspent_joint_output = false;
             }
             OracleAttestedPriorCetTimelock {
                 timelocked_cet,
```
Comment (bonomat) [inline daemon/src/projection.rs:504]: 📯
Is it correct that we can achieve this if we publish the commit transaction too late and the CET timelock has not expired yet when the oracle attests to the price? 

I think this needs to be false as well. If the oracle attests to the price we know already the closing price.

**Review by bonomat — CHANGES_REQUESTED** [review]
Thanks. I think the `OracleAttestedPriorCetTimelock` should also set unpsnet_joint_output to false

File: daemon/src/projection.rs (hunk 2 above)
Comment (thomaseizinger) [reply daemon/src/projection.rs:504]: Hmm, interesting!

I guess we can say that having a valid but _timelocked_ transaction that can spend the joint output should also already account for the output being spent (which is what the boolean represents).

Comment (thomaseizinger) [reply daemon/src/projection.rs:504]: > I think this needs to be false as well. If the oracle attests to the price we know already the closing price.

By itself, that reasoning does not make sense though. This boolean tracks whether we have an unspent join output, not whether we know the closing price. I think this comment highlights that we are introducing an unfit abstraction here.

**Comment by da-kami** [comment]
Thanks for the fix, I think we can merge this!

**Comment by thomaseizinger** [comment]
> Thanks for the fix, I think we can merge this!

Yeah. It think it should be fine although the more I think about this, the less I like the abstraction we are introducing. It is such a mess :(

**Comment by da-kami** [comment]
> > Thanks for the fix, I think we can merge this!
> 
> Yeah. It think it should be fine although the more I think about this, the less I like the abstraction we are introducing. It is such a mess :(

Because the Boolean is not representative enough? 
We can introduce more "states" - this goes back to the discussion what this endpoint should be used for. "Open" is an ambiguous state, for the automation it matters if we have the closing price or not to determine "open". If we agree that this endpoint is for the automation only the current representation is sufficient.

**Comment by thomaseizinger** [comment]
> > > Thanks for the fix, I think we can merge this!
> > 
> > Yeah. It think it should be fine although the more I think about this, the less I like the abstraction we are introducing. It is such a mess :(
> 
> Because the Boolean is not representative enough? 
> We can introduce more "states" - this goes back to the discussion what this endpoint should be used for. "Open" is an ambiguous state, for the automation it matters if we have the closing price or not to determine "open". If we agree that this endpoint is for the automation only the current representation is sufficient.

You are missing the point. It is not about more or less states. Also, open is not ambiguous. Unless we operate on randomness, there is no ambiguity here!

The problem is that we are defining an abstraction (open) based on needs (automation) that the current software (maker) is not and cannot be aware of.

If the maker should exist as an independent piece of software, the abstractions that are provided should make sense in isolation, otherwise they are bad abstractions.

In its current state, there is no decision logic in the maker. Consequently, any abstraction that is introduced needs to make sense in isolation without knowledge of what the APIs are used for.

The only reasonable abstraction I could come up with is to define "open" around the blockchain state of the CFD: whether the joint output is spent or not. This has nothing to do with the price.

The fact that we need to slightly skew this abstraction and consider "joint output will likely be spent in a few blocks" as "has no joint output and is thus not open anymore" is what I call a mess.

I suggest to read https://github.com/itchysats/architecture/pull/25 for some more context!

**Review by bonomat — APPROVED** [review]
thanks

**Comment by da-kami** [comment]
> [snip]
> You are missing the point. It is not about more or less states. Also, open is not ambiguous. Unless we operate on randomness, there is no ambiguity here!
> 
> The problem is that we are defining an abstraction (open) based on needs (automation) that the current software (maker) is not and cannot be aware of.

For me it was not the point of this PR to invest time into re-designing the API, but to fix a problem where the automation reacts to a state change that should not be (the scenario where a contract setup fails, we treat `ContractSetup` as open in the automation because rollover uses the same state...).

What are you aiming for with this PR? Fixing this problem? Re-design?

> [snip]
> I suggest to read [itchysats/architecture#25](https://github.com/itchysats/architecture/pull/25) for some more context!

I read through it, will react there. I am not sure we have a decision there yet and it will take time to implement this. Given what we currently have it would be great to land some "messy" fix in the meanwhile. That does not have to be an `open` endpoint, we could still opt for using a different state when rolling over.

**Comment by luckysori** [comment]
Let's close this PR for the time being, since we already introduced a quick fix with https://github.com/itchysats/itchysats/pull/1780.
